### ISSUE TITLE
impossible to see the existings shortcut + black and white steam games icon

### ISSUE THREAD

**Issue by sk666** [terrymacdonald/DisplayMagician#22]
**Describe the bug**
Today, with the latest version of display magician 1.1.4.0, I am unable to see my existing game shortcut in the software, as you can see on the pictures. Another "problem" is the icons of the steam games are now black and white, i don't now if it's a problem or if it's normal. Before the 1.1.1.0, the icons was in colors.

**To Reproduce**
Steps to reproduce the behaviour:
1. Click on 'game shortcut in the main window'
2. See error

**Expected behaviour**
Normally, I should see ATS surround shortcut previously created. The shortcut is still on my desktop but impossible to see it in the software. I can edit it because the shortcut exist but I don,t see it.

**Screenshots**
![screen 1](https://user-images.githubusercontent.com/29050341/123523389-3acc4680-d691-11eb-918a-20f07d86855a.PNG)
![screen 2 error (french)](https://user-images.githubusercontent.com/29050341/123523391-3b64dd00-d691-11eb-97d0-d69e86faabe7.PNG)

**Reporting Logs**
If required I can add. Please let me know if they are needed.

**Enviroment (please complete the following information):**
 - Windows Version: Win10 21H1
 - DisplayMagician Version 1.1.4.0
 - Video Card: NVIDIA Geforce GTX 970
 - Number and make/model of monitors: 2xDell 2208WFP
 - Date of last video driver update: 2021-05-17 (driver version 466.63)
 - Date of last windows update: 2021-06-17

**Additional context**
The last things I did before this problem happening was an edit of the shortcut in this section, so I presume I did something wrong. Concerning the black and white icons, this problem happening just vefore an update of the software, may be in april or may 2021.

I'm very sorry for my english, it's not my first language. Please let me know if I can something for you to help me about theses issues.

**Comment by terrymacdonald** [comment]
Hi @sk666,

So firstly, check the shortcut icons listed in the Shortcut Library. Black and white shortcut icons within the shortcut library and with an exclamation symbol on them indicate that there is a problem with the shortcut that will prevent it from running. This could be anything from the display profile used by the game shortcut no longer being correct, to the game being uninstalled from the game library (e.g. uninstalling the game from within Steam).

I will need the logs in order to troubleshoot this. This is an unusual error and one I've not seen before.

I will need to see a TRACE DisplayMagician.log file, and I need a copy of your Shortcuts_1.0.json file and your DisplayProfiles_1.0.json file in order to troubleshoot this.

You can create a TRACE DisplayMagician.log by opening a command prompt, and running "C:\Program Files\DisplayMagician\DisplayMagician --trace". This command will make DisplayMagician generate a DisplayMagician.log file in the C:\Users\<yourusername>\AppData\Local\DisplayMagician\Logs folder. Please go there, open the DisplayMagician.log file and check that there are a lot of 'TRACE' log entries in the log file. If there aren't, then you'll need to try to run the command again to make DisplayMagician generate the DisplayMagician.log file.

The Shortcuts_1.0.json file will be in the C:\Users\<yourusername>\AppData\Local\DisplayMagician\Shortcuts folder, and the DisplayProfiles_1.0.json file will be in the C:\Users\<yourusername>\AppData\Local\DisplayMagician\Profiles folder. I'll need both of them in order to troubleshoot this.

Did you manually edit Shortcuts_1.0.json or DisplayProfiles_1.0.json before you had the issue? Or did you edit them through the DisplayMagician program itself?

BTW Your English is amazing @sk666 - way, way better than my French! You don't need to apologise at all.

Thanks
Terry

**Comment by sk666** [comment]
Hi Terry,

thank you for the fast reply! Concerning the error message, I found the problem yesterday, a couple hour after posting this but before your reply, I need to say that the missing shortcut in display magician works even if I don't see it in the "game shortcut library" inside Display magician (I was able to launch it with the copy of the shortcut on my desktop, I saw the switch to the surrund profile like it should be so it's very strange haha).

So like I said, the error message was my fault, I edited the shortcut inside display magician itself before this problem happens. So I manually serach on my computer to find if the shortcut was stored in the registry or in the app data, found it in appdata and I opened it to see if I was able to understand the script. I understood how the file works so I was confident to cut my last edit inside the file itself, in fact this part: ![cut this to resolve the error message problem](https://user-images.githubusercontent.com/29050341/123541943-f08ca900-d714-11eb-8bdc-d611a4c70478.PNG)) and voila, the problem was solved with this simple way! I am able to see now my previously created shortcut in display magician again :)

I made a copy of the shortcut.js before I fixed it, let me know if you still want it (and if you still want the other logs), because the error message problem dosen't exist anymore. I was able to fix it but honestly, I'm not sure if I understand what happens, I put a shortcut of the Steam game on my desktop, I think my error was that I pointed the executable on this game shortcut instead of the steam game itself but I'm not sure if this was the cause of the problem. 

For the black and white icon (may be I should have created 2 issue for the 2 problem insted of 1 issue for both problem), in fact I don't have explanation symbol on them. I see what your talking about but I'm sure that is not that. The explanation symbol may appear when you delete a profile linked to a previously created shortcut linked to this profile. When this happens, display magician warns you that there something wrong with the profile itself and you need to link it again in the shortcut of the game shortcut library section. I need to specify that the shortcut in display magician works with the black and white icon, The steam games icon are just black and white and this happend just after a previously update of display magician. I did not try to downgrade the display magician version, I don't want to break something again. I see the colored icon for just 2 steam games on my library as you can see in this picture: 

![black and white icons exept 2 of them](https://user-images.githubusercontent.com/29050341/123542621-9ee61d80-d718-11eb-8732-b40e4f988bec.PNG) 

and this picture too: 

![black and white without explanation symbol](https://user-images.githubusercontent.com/29050341/123542924-297b4c80-d71a-11eb-8e6f-72db212aea6e.PNG)


The black and white icons let me launch the game and display magician works but I had a couple of weird things after the black and white icons problem happens. Display magician seems to be instable since this time but i'm not sure if it reliable to the black and white icon itself. I saw that you did a couple update of display magician in the past month, so may be it will help alot. I don't know if the black and white icons are "normal" or not. Let me know if you want something for the black and white icons problem.

thank you again for your work, display magician is better than Nvidia surround itself! Sorry for the long post :s

**Comment by terrymacdonald** [comment]
Hi,

I will need the TRACE DisplayMagician.log file, and the shortcut and
displayprofile JSON files to troubleshoot that error. As mentioned, I've
never seen that issue before, so I want to make sure that I really
understand what is happening.

Thank you so much for such a good explanation. Having that much detail
really helps me understand what is happening. I will still need those other
files though, as I need to know exactly which bit of the code is failing,
and those log files and JSON files will help with that.

Thanks
Terry

**Comment by sk666** [comment]
Hi Terry,

allright, that are the log and JSOn files taht you need. Please let me know if I can do something for you.

[logs.zip](https://github.com/terrymacdonald/DisplayMagician/files/6726447/logs.zip)

**Comment by terrymacdonald** [comment]
Hi @sk666 ,

Thanks for the files. 

Firstly I can see that you've directly added American Truck Simulator as an Executable. I would suggest changing that shortcut so that you use the 'Launch a Game installed in Steam or Uplay' option, and select your game from the list there. 

![image](https://user-images.githubusercontent.com/5721772/123700595-cb36a100-d8b4-11eb-889b-2d3475082ab8.png)

I *think* that the executable option may be the reason the icon is grayscale rather than full colour, but I'll need some TRACE log entries to be sure.

NOTE: The DisplayMagician.log file is unfortunately just the default DisplayMagician.log file. If you open it you'll see that it only has 'WARN' entries in it. I need the TRACE entries in it too.

To generate a DisplayMagician.log file with TRACE entries in it, please do the following:

- Press WindowsKey + R to bring up a Run prompt
- Click the Browse button
- Navigate to C:\Program Files\DisplayMagician\DisplayMagician.exe, and click the Open button
- Go to the end of the command, and add a space, and then add `--trace` on the end.
- The whole command should read `"C:\Program Files\DisplayMagician\DisplayMagician.exe" --trace`
- Press the Run button.

This will force DisplayMagician to start logging TRACE commands until DIsplayMagician is closed down.

Now. perform the actions that don't work. In your case, please open up the DisplayProfile window, and then close it down, and then open up the Shortcut Library and then edit a Shortcut (but don't change anything). Then close down DisplayMagician.

Then straight away please zip up the DisplayMagician.log file and upload it here again. Please make sure that it has TRACE commands in the log file, because then we know that the --trace option worked properly.

Once I have the DisplayMagician.log file with TRACE commands in it I'll be able to see if there is anything wrong.

**Comment by sk666** [comment]
Hi Terry,

I think you did not understand what I tried to said earlier. All problem are solved except for the black and white icons. I have sended you the old log that contains the previously problem of missing shortcut, I sended you the old log in case that you want to see the problem that I had. This problem is not present anymore, I fixed it earlier, as you can see here:
![shortcut not appear solved](https://user-images.githubusercontent.com/29050341/123807649-0be6e800-d8be-11eb-80ac-f111627feb2d.PNG)


So for the black and white icons, I can send you the trace log, but I already have selected launch from Steam, as you can see here:
![launch from steam](https://user-images.githubusercontent.com/29050341/123807707-1903d700-d8be-11eb-89a8-8f35498b08f8.PNG)

I tried to generate the trace log, is it the right log?: 
[DisplayMagician.log](https://github.com/terrymacdonald/DisplayMagician/files/6733841/DisplayMagician.log)

Thank you!

**Comment by terrymacdonald** [comment]
Hi Sk666

I think I did understand, but I thought that there could be an underlying
issue with how it was grabbing the bitmaps.

Thank you for the TRACE DisplayMagician.log file. It helped me validate
that the program appears to be working properly (at least as far as I can
tell).

The black and white icons may actually be the icons chosen from the
selected icon file. As an example, DisplayMagician chose to get the first
256x256 it found within
the D:\Jeux\PC\STEAM\steam\games\5daee883eed0789df7a96d8ffe89def35edc1431.ico
file as the picture it wanted to use for Euro Truck Simulator 2. It may be
that the first 256x256 icon within
the D:\Jeux\PC\STEAM\steam\games\5daee883eed0789df7a96d8ffe89def35edc1431.ico
file is actually black and white.

It sounds like DM is working for fine now, so maybe we'll just leave things
there. If you did want to confirm whether Euro Truck Simulator 2 has a
black and white icon in it you can use these instructions
<https://www.tenforums.com/tutorials/128170-extract-icon-file-windows.html> but
don't feel like you have to do this. I'm happy to close this issue now if
you are also happy for me to do that.

Thanks
Terry

**Comment by terrymacdonald** [comment]
Closing as problem seems to be solved. Feel free to reopen if the issue still remains.

**Comment by sk666** [comment]
Hi Terry,

I think your right, I checked out inside the .ico with the software Iconviewer, the first one ico inside is always a black and white picture. I tried it just for the fun to extract and convert them again in .ico (it take less than 3 minutes to do it) and it works like you said as you can see here:
![colored icons DM](https://user-images.githubusercontent.com/29050341/123878364-f0ef9480-d90c-11eb-9e2f-afafabe74d8c.PNG). 

Now, DM display the colored ico like he did in a previous update, I saw that the colored 256x256 32 bits (PNG) ico inside the original steam ico are always the last one 256x256 32 bits (PNG), I don't know if it's possible to you to make it choose the last one 256x256 32 bits (PNG) instead the first one white and black. If it's possible, I think it will fix the "bug". Feel free to close this issue, even if you dosen't want to fix this problem. I just wanted to know why some icons are colored and why the other ones are black and white, I understand it now with your explanations.

Thank you again for all your help and your time :)

**Comment by terrymacdonald** [comment]
Hi Sk666,

I purposely chose the first matching 256x256 icon as most modern games have
the main icon as the first icon in the list, generally in increasing size
from 32x32 up to 256x256. I tested many games to find a solution that would
extract the best quality icon images as possible. I found that game makers
sometimes include additional sized icons later on in the icon file to allow
people to swap if they want, or for other purposes, but in general the
first series of icons in the icon file is the one that is the main one.

It looks as if the truck simulator games have decided to use the reverse
structure to most other games I've tested. I don't think it's best to
reverse the icon fetching logic as it will affect the other games too much
in order to fix these games. This is the first example of this reverse icon
structure I've seen, so my recommendation is that we stick with the
existing logic as it is, and just hope that games designers choose the more
common way of doing things :)

Or, of course, people can manually make the changes as you have done. Nice
work BTW!

Thanks
Terry
